### DISCUSSION TITLE
Random chats

### DISCUSSION THREAD

**Discussion by ogamespec** [emu-russia/dmgcpu#131] [General]
I'm going to put all kinds of discord chats with valuable information here. If not here, where? :)

```
@gekkio
  -- ADON
  a_reg_hilo_gte_8 <= a_reg(7) and a_reg(4);
  -- ALYP
  a_reg_hi_gt_9 <= a_reg(7) and (a_reg(6) or a_reg(5));
  -- AHEE
  a_reg_lo_gt_9 <= a_reg(3) and (a_reg(2) or a_reg(1));


your bq7 is the same as my a_reg_hilo_gte_8, which means "are both nibbles greater than or equal to 8"
your bq5 is the same as my a_reg_hi_gt_9, which means "is the upper nibble greater than 9"
your bq4 is the same as my a_reg_lo_gt_9, which means "is the lower nibble greater than 9"
those are used by the ALU to generate an "adjustment operand" that gets added to A during DAA
```

**Comment by ogamespec** [comment]
```
@gekkio
the "bc" circuit has two slightly separate parts: one part outputs the ALU flags to one of the buses during PUSH AF. The other one generates an ALU operand for one step of the SP+e instructions, but I'm not 100% sure my analysis is correct since I haven't yet studied it in detail
```

**Reply by ogamespec** [reply]
Regarding "bc". I tend to think that even though bc0/bc4 is at the bottom, it is still part of the ALU. I'll probably move this circuit in my HDL inside the ALU instead of at the bottom. Then `wire [5:0] bc;` will become `output` instead of `inout` as it is now.

**Comment by ogamespec** [comment]
@Sono
lmao, I think the unbonded input is the NMI
![image](https://user-images.githubusercontent.com/5828819/197352647-c2862f92-b985-4fa7-8f6f-6697b2ec8221.png)
it's glitchy though
yup, it has to be NMI 100%, because switching the NMI pin again doesn't re-trigger

**Reply by ogamespec** [reply]
@gekkio
yes, that's correct. It even has the same input structure in the SoC as the RES pin that probably contains a schmitt trigger to reduce sensitivity to noise

**Reply by msinger** [reply]
I just randomly stumbled on this here... NMI and RES are not the same. NMI has a pull-up resistor, implemented by an always enabled PMOS with a longer gate. RES doesn't have a pull-up resistor. NMI needs the pull-up, because it is not bonded to an external pin, so it would float otherwise. I believed they were identical at first too. But because I didn't see the pull-up and knew it can't possibly be floating, I was bothered by this for a long time, so I stared at the die shot for a very long time and finally I saw the difference between them. :)

I documented them as [IBUFFER - Variant A](http://iceboy.a-singer.de/doc/dmg_cells.html#ibuf_a) (RES) and [IBUFFER - Variant B](http://iceboy.a-singer.de/doc/dmg_cells.html#ibuf_b) (NMI) already.

Other than the pull-up and some diodes, they are just an inverter with differently sized PMOS and NMOS transistors.

**Reply by ogamespec** [reply]
Thanks for the info, I wonder why they didn't put /NMI outside? It would be useful from the cartridge to be able to trigger /NMI.

**Reply by msinger** [reply]
Yes, they could have ditched that stupid analog pin on the cartridge and replace it with the NMI. I'm not sure if any game ever used that analog input or if it was only used by some "homebrew" cartridges.

The CGB chip has the NMI bonded to an external pin btw. You can see it [here](https://gamesx.com/wiki/lib/exe/fetch.php?media=schematics:gameboy_color_service_manual.pdf) at the very last page. Pin 57. But it just goes to a test point on the PCB and is maybe connected to VDD. Nothing else.
